src/Lib/BSP/IO/LumpReader.cs

```diff
@@ -0,0 +1,101 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using SharpCompress.Compressors.LZMA;
+using Lumper.Lib.BSP.Lumps;
+
+namespace Lumper.Lib.BSP.IO
+{
+    // handles decompressing and fills lumps with data
+    public abstract class LumpReader : BinaryReader
+    {
+        // lumpheader information is only needed in the reader
+        protected List<Tuple<Lump, LumpHeader>> Lumps = new();
+        public LumpReader(Stream input) : base(input)
+        { }
+        protected abstract void ReadHeader();
+        protected virtual void LoadAll()
+        {
+            foreach (var l in Lumps)
+            {
+                var lump = l.Item1;
+                var lumpHeader = l.Item2;
+                if (lumpHeader.Length > 0)
+                    Read(lump, lumpHeader);
+            }
+        }
+        private MemoryStream Decompress()
+        {
+            MemoryStream decompressedStream = new();
+
+            const string magic = "LZMA";
+            var id = ReadBytes(magic.Length);
+            if (Encoding.ASCII.GetString(id) != magic)
+                throw new InvalidDataException("Failed to decompress: Lump doesn't look like it was LZMA compressed");
+
+            uint actualSize = ReadUInt32();
+            uint lzmaSize = ReadUInt32();
+            byte[] properties = new byte[5];
+
+            Read(properties, 0, 5);
+
+            var lzmaStream = new LzmaStream(properties,
+                                            BaseStream,
+                                            lzmaSize,
+                                            actualSize);
+            lzmaStream.CopyTo(decompressedStream);
+            decompressedStream.Flush();
+            decompressedStream.Seek(0, SeekOrigin.Begin);
+
+            return decompressedStream;
+        }
+        protected void Read(Lump lump, LumpHeader lumpHeader)
+        {
+            var decompress = lump is not IUnmanagedLump;
+
+            BinaryReader lumpReader;
+            long lumpStreamLength;
+
+            BaseStream.Seek(lumpHeader.Offset, SeekOrigin.Begin);
+
+            if (lumpHeader.Compressed)
+            {
+                if (decompress)
+                {
+                    MemoryStream decompressedStream = Decompress();
+                    lumpReader = new BinaryReader(decompressedStream);
+                    lumpStreamLength = decompressedStream.Length;
+                    if (lump is IUnmanagedLump unmanagedLump)
+                    {
+                        unmanagedLump.Compressed = false;
+                        unmanagedLump.UncompressedLength = -1;
+                    }
```

Comment:
If `decompress` is true then `lump` is not `IUnmanagedLump`. So this is always false

Reply:
Yes but only because "decompress" is hard coded to check the type right now. Not sure why we would need to decompress a unmanaged lump in the future and not make it a managed type instead but in theory you can decompress them lumps as well so that's why its here. I would keep it.

Reply:
If it could be useful in the future then you could comment it out but leave it, but I don't think it's right to have code that's logically impossible to be true in uncommented code. At the very least leave a comment to say what you just said, we don't want other programmers reading the code left scratching their heads over blocks of code that make no sense logically